Refactor `preferences`

not-my-profile opened the pull request:
## What does this PR do?

Fixe some hacks & misplaced inheritance and adds type hints.

## Why is this change important?

<!-- MANDATORY -->

<!-- explain the motivation behind your PR -->

## How to test this PR locally?

<!-- commands to run the tests or instructions to test the changes-->

## Author's checklist

<!-- additional notes for reviewiers -->

## Related issues

<!--
Closes #234
-->

not-my-profile committed:
[refactor] add type hints & remove Setting._post_init

Previously the Setting classes used a horrible _post_init
hack that prevented proper type checking.

not-my-profile committed:
[refactor] remove pointless tuple

not-my-profile committed:
[refactor] stop SwitchableSetting from subclassing Setting

Previously the default_value was abused for the cookie name.
Having SwitchableSetting subclass Setting doesn't even make sense
in the first place since none of the Setting methods apply.

return42 reviewed (commented):
Wow another gorgeous work from you .. thanks!

I just want to leave some comments from my review .. and before we merge I have a question, see below ..

----

from the both commits:

> Previously the Setting classes used a horrible _post_init hack that prevented proper type checking.
> [refactor] remove pointless tuple

Oh yes .. very old artefact from fe691a09 / thanks for refactoring this.

> [refactor] rename SwitchableSetting to FeatureSet

I think the name 'SwitchableSetting' is more precise here (even if that name sounds crude to me).

The *Switchable* indicates a on/off (enable/disable) state of such a setting.  Same with the names 'Choice' vs. 'Feature' .. I don't really like 'SwitchableSetting' and 'Choice' but I think 'Feature' and 'FeatureSet' sounds too generall in my ears.

This is just my opinion ... What do you think?

> [refactor] stop FeatureSet from subclassing Setting

I'm wondering about the old implementation (added in 99435381) .. now, when I look at, the old implementions looks crude to me. / thanks!

return42 commented on searx/preferences.py:
```diff
@@ -200,19 +200,19 @@ def save(self, name: str, resp: flask.Response):
-    """Base class for settings that can be turned on && off"""
+class FeatureSet(Setting):
+    """A set of features that can be either enabled or disabled."""
 
```
What about `Choice` --> `Toggle` and `SwitchableSetting` --> `ToggleSet`?

not-my-profile committed:
[refactor] refactor SwitchableSetting

The previous implementation used two hash sets and a list.
... that's not necessary ... a single hash map suffices.

And it's also less error prone ... because the previous data structure
allowed a setting to be enabled and disabled at the same time.

return42 commented on searx/preferences.py:
```diff
@@ -185,34 +177,34 @@ def parse(self, data):
-        if not hasattr(self, 'map'):
-            raise MissingArgumentException('missing argument: map')
-        if self.value not in self.map.values():  # pylint: disable=no-member
+    def __init__(self, default_value, map: Dict[str, object], locked=False):  # pylint: disable=redefined-builtin
```
sorry for bikesheeding again ...

`map` is a ~keyword~ builtin .. I'm often at the same situation where a name that really fits is a keyword or a builtin  e.g. `map` or `dict` .. I usually then choose a name like `kv` (key/value) ... `kvmap` ..

Here I would decide to simply use `values` .. what also fits to the usage:

![grafik](https://user-images.githubusercontent.com/554536/148514777-22cff3fb-496a-4234-bb30-50e83ba803df.png)

compare with ...

![grafik](https://user-images.githubusercontent.com/554536/148514335-dacaaf16-9dc8-445a-be66-1e55fc251a75.png)


But it is a nit .. if you think `map` is good enough leave it as is.

not-my-profile replied:
nit: `map` isn't a keyword ... it's just a built-in function (otherwise this would be a syntax error).

Yeah I'm currently not really interested in fixing such pylint errors.

return42 replied:
> nit: map isn't a keyword ... it's just a built-in function (otherwise this would be a syntax error).

right. Sorry I'm fighting with my github editor .. I edited my initial comment ..

not-my-profile replied:
>map is a keyword

[No it's not.](https://docs.python.org/3/reference/lexical_analysis.html#keywords) as I explained it's a [built-in function](https://docs.python.org/3/library/functions.html).

The difference is that `map = 1` works while e.g. `for = 1` is a syntax error.

(But yes I know it's bad practice to shadow built-ins ... it's just not important enough for me to fix right now)

return42 reviewed (approved):
Thanks for your clean up!

`map` is OK for me .. but `values` fits IMO better when we look at the usage of this argument .. and is not a builtin.

But as I said, it is just bikesheeding.
